"Graceful" Server Termination #188

ktoso opened the pull request:
Resolves: https://github.com/akka/akka-http/issues/188
Relates to https://github.com/akka/akka/pull/18475

Still some pondering to do in the terminating phase in the terminator, but generally seems correct.
I also will clean up documentation, but please review already

ktoso commented on akka-http-core/src/main/resources/reference.conf:
```diff
@@ -155,6 +155,18 @@ akka.http {
+    # TODO currently not possible to make that response created/evaluated once the termination triggers this could allow "rebalance to X.Y.Z.Z " responses
+    termination-deadline-exceeded-response {
+      # Status code of the "terminating" response to be automatically sent to pending requests once the termination deadline is exceeded.
+      status = 503 # ServiceUnavailable
```
TODO make a test for it

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,294 @@
+/**
+ * Used to start the termination process of an Akka HTTP server.
+ *
+ * TODO more details here?
```
TODO

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,294 @@
+}
+
+// TODO message better?
+final class ServerTerminationDeadlineReached() extends RuntimeException("Server termination deadline reached, shutting down all connections and terminating server...")
```
message does not really matter too much i think... since we handle it internally only, will remove todo

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,294 @@
+
+object GracefulTerminatorStage {
+  def apply(system: ActorSystem, serverSettings: ServerSettings): BidiFlow[HttpResponse, HttpResponse, HttpRequest, HttpRequest, ServerTerminator] = {
+    val enable = true // FIXME get from settings?
```
WDYT, setting or not? I guess not

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,294 @@
+ * toNet   Response  <- | S | <- fromUser Response
+ *                      +---+
+ * }}}
+ */
```
the most detailed and up to date writeup

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,294 @@
+      })
+
+      // TODO what other semantics should we change in this phase...
+      def installTerminationHandlers(deadline: Deadline): Unit = {
```
some semantics debate could happen in those handlers... those seem correct AFAICS but maybe more tests needed

ktoso commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -161,14 +160,24 @@ class HttpExt private[http] (private val config: Config)(implicit val system: Ex
+          ServerBinding(tcpBinding.localAddress)(
+            () ⇒ tcpBinding.unbind(),
+            deadline ⇒ {
+              ??? // FIXME uhm... do the same terminator dance here I guess
```
FIXing this now, sadly a bit of setup dance has to be replicated from the other place

ktoso commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -70,8 +72,8 @@ class HttpExt private[http] (private val config: Config)(implicit val system: Ex
-  private type ServerLayerBidiFlow = BidiFlow[HttpResponse, ByteString, ByteString, HttpRequest, NotUsed]
-  private type ServerLayerFlow = Flow[ByteString, ByteString, Future[Done]]
+  private type ServerLayerBidiFlow = BidiFlow[HttpResponse, ByteString, ByteString, HttpRequest, ServerTerminator]
+  private type ServerLayerFlow = Flow[ByteString, ByteString, (Future[Done], ServerTerminator)]
```
the type is never exposed completely to users, so i'm not worried about changing it

ktoso commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -86,7 +88,7 @@ class HttpExt private[http] (private val config: Config)(implicit val system: Ex
       }
 
-    serverBidiFlow
+    GracefulTerminatorStage(system, settings) atop serverBidiFlow
```
this is where we instal the logic

raboof replied:
Do we have any feeling for how much overhead this feature causes? I guess if it's noticeable we could make it configurable so people for whom unbind+timeout is sufficient can opt out of it?

ktoso replied:
I think not much... it is a "pass through" most of the time, which should be fused

ktoso replied:
And while terminating it should not matter that much since goal is to quickly send 500s anyway

ktoso replied:
I have not yet benchmarked though

ktoso commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -195,14 +204,23 @@ class HttpExt private[http] (private val config: Config)(implicit val system: Ex
+      fuseServerFlow(fuseServerBidiFlow(settings, connectionContext, log), handler)
+
+    // "T-800 101" - the "one" terminator that we emit to user land, while other terminators are materialized per-connection.
+    val masterTerminator = new MasterServerTerminator(log)
```
meh removing that joke btw

raboof commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,289 @@
+
+  final val TerminationDeadlineTimerKey = "TerminationDeadlineTimerKey"
+
+  final class ConnectionTerminator(triggerTermination: Promise[FiniteDuration ⇒ Future[HttpTerminated]]) extends ServerTerminator {
```
Pretty intricate logic here.. can be at least a little simpler by taking a `triggerTermination: Future[FiniteDuration ⇒ Future[HttpTerminated]]` here

raboof commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,289 @@
+
+      // this promise will be completed once our termination is complete;
+      // e.g. we replied with "go away" to pending requests, and no new ones were incoming etc
+      val terminationOfConnectionDone = Promise[HttpTerminated]() // FIXME this is a wrong name... it is only the connection that terminated here
```
looks like the FIXME can be removed since the variable was already renamed?

raboof commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,289 @@
+        case TerminationDeadlineTimerKey ⇒
+          val ex = new ServerTerminationDeadlineReached
+          if (pendingResponse) {
+            // sending the reply here is a "nice to try", but the stage failure will likely overtake it and terminate the connection first
```
Not sure if worth it, perhaps OK to just fail/close?

raboof commented on akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala:
```diff
@@ -26,5 +30,46 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
+
+  // TODO docs after review
+  def terminate(hardDeadline: java.time.Duration): CompletionStage[HttpTerminated] = {
+    delegate.terminate(FiniteDuration.apply(hardDeadline.toMillis, TimeUnit.MILLISECONDS))
```
didn't we have a `toScala` somewhere for JDuration->FiniteDuration?

ktoso replied:
thanks!

raboof commented on akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala:
```diff
@@ -26,5 +30,46 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
+
+  /**
+   * Completes when the [[terminate]] is called and server termination is in progress.
+   * Can be useful to make parts of your application aware that termination has been issued,
```
Maybe also explain relationship with coordinated shutdown also here?

ktoso replied:
Added it

raboof commented on akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala:
```diff
@@ -26,5 +30,46 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
+   * This signal can for example be used to safely terminate the underlying ActorSystem.
+   *
+   * Note: This mechanism is currently NOT hooked into the Coordinated Shutdown mechanisms of Akka.
+   *       TODO: This feature request is tracked by: https://github.com/akka/akka-http/issues/1210
```
:+1:

ktoso commented on akka-http-core/src/main/mima-filters/10.1.2.backwards.excludes:
```diff
@@ -3,4 +3,11 @@ ProblemFilters.exclude[Problem]("akka.http.impl.*")
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.Http#ServerBinding.copy")
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.Http#ServerBinding.this")
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.http.scaladsl.Http#ServerBinding.apply")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.settings.ServerSettings.terminationDeadlineExceededResponse")
```
oh weird, was copy pasted from mima, seems it reported the thingy twice

ktoso commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -883,24 +910,94 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
    *
    */
-  final case class ServerBinding(localAddress: InetSocketAddress)(private val unbindAction: () ⇒ Future[Unit]) {
+  final case class ServerBinding(localAddress: InetSocketAddress)(
```
Hmm I wonder... you think we should?

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -451,6 +451,11 @@ private[http] object HttpServerBluePrint {
 
+            case _: ServerTerminationDeadlineReached ⇒
+              val response = settings.terminationDeadlineExceededResponse
+              log.warning(s"Closing connection with [${response.status}] response, termination deadline exceeded. Server is shutting down...")
```
use logging placeholder

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,296 @@
+      case v @ AliveConnectionTerminators(emptyTs) if emptyTs.isEmpty ⇒
+        if (terminators.compareAndSet(v, Terminating(timeout.fromNow))) {
+          // no connections exist to be terminated, and if some arrive after us, they will see the Terminating and terminate
+          termination.success(HttpServerTerminated)
```
sure that it's safe? relation to below `termination.completeWith`? otherwise `trySuccess`

ktoso replied:
Yeah I think it's safe due to the CAS protecting it, but won't hurt to make it a trySuccess I think

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,296 @@
+}
+
+final class ServerTerminationDeadlineReached()
+  extends RuntimeException("Server termination deadline reached, shutting down all connections and terminating server...")
```
is the stack trace interesting? will it be logged?

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,296 @@
+ */
+// "Hasta la vista, baby."
+@DoNotInherit
+abstract class ServerTerminator {
```
this is in `impl` package, is it still public API? then perhaps in wrong package?
otherwise InternalApi instead of DoNotInherit

ktoso replied:
Yes public API, I will move it

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,296 @@
+  }
+}
+
+final class ServerTerminationDeadlineReached()
```
public api vs `impl` package?

patriknw replied:
then I think I'd prefer to mark it as internal rather than having grey zones

ktoso replied:
Ok

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,296 @@
+final class ServerTerminationDeadlineReached()
+  extends RuntimeException("Server termination deadline reached, shutting down all connections and terminating server...")
+
+object GracefulTerminatorStage {
```
internal

ktoso replied:
yes thanks

patriknw replied:
still not internal

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,296 @@
+ *                      +---+
+ * }}}
+ */
+final class GracefulTerminatorStage(settings: ServerSettings)
```
internal

ktoso replied:
yes, thanks

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,296 @@
+    // responsible for terminating this connection
+    val selfTerminator = new ConnectionTerminator(triggerTermination)
+
+    val logic = new TimerGraphStageLogic(shape) with StageLogging {
```
good to override `logSource` for better logger name

ktoso replied:
good :)

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/settings/ServerSettingsImpl.scala:
```diff
@@ -100,5 +101,15 @@ private[http] object ServerSettingsImpl extends SettingsCompanion[ServerSettings
+  private def terminationDeadlineExceededResponseFrom(c: Config): HttpResponse = {
+    HttpResponse(
+      status = StatusCodes.getForKey(c getInt "termination-deadline-exceeded-response.status")
+        .getOrElse(throw new IllegalArgumentException("Illegal status code set for `termination-deadline-exceeded-response.status`!"))
```
include the configured value?

ktoso replied:
thanks

patriknw commented on akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala:
```diff
@@ -26,5 +32,84 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
+      .toJava
+
+  /**
+   * This future completes when the termination process, as initiated by an [[terminate]] call has completed.
```
s/future/CompletionStage/

ktoso replied:
thanks

patriknw commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -217,7 +234,15 @@ class HttpExt private[http] (private val config: Config)(implicit val system: Ex
+              () ⇒ tcpBinding.unbind(),
+              timeout ⇒ masterTerminator.terminate(timeout)(fm.executionContext)
+            )
+        )(ExecutionContexts.sameThreadExecutionContext)
```
are you 110% sure that will never throw?

patriknw commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -883,24 +910,124 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
+     * @return future which completes successfully with a marker object once all connections have been terminated
+     */
+    def terminate(hardDeadline: FiniteDuration): Future[HttpTerminated] = {
+      require(hardDeadline.toMillis > 0, "deadline must be greater than 0ms")
```
`hardDeadline > Duration.Zero`

patriknw commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -883,24 +910,124 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
+      require(hardDeadline.toMillis > 0, "deadline must be greater than 0ms")
+
+      _whenTerminationSignalIssued.trySuccess(hardDeadline.fromNow)
+      val terminated = unbindAction().flatMap(_ ⇒ terminateAction(hardDeadline))(ExecutionContexts.sameThreadExecutionContext)
```
sure it will never throw?

ktoso replied:
I think unbind is more possible to throw perhaps... I'll require an execution context then

ktoso commented on akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala:
```diff
@@ -26,5 +32,84 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
+   * @return completion stage which completes successfully with a marker object once all connections have been terminated
+   */
+
+  def terminate(hardDeadline: java.time.Duration): CompletionStage[HttpTerminated] = {
```
I was thjinking hardDeadline explains more that "THIS WILL KILL CONNECTIONS" :-) 
WDYT, shall I rename? I don't mind too much

ktoso commented:
Pushed a bit more, had one or two questions in the comments... will squash and finish tomorrow morning

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,293 @@
+
+            // on purpose discard all incoming bytes for requests
+            // could discard with the deadline.timeLeft completion timeout, but not necessarily needed
+            request.entity.discardBytes()(interpreter.subFusingMaterializer)
```
I'm not entirely sure about this... I think it's fine but have not written a test for it yet...
I mean, this is effectively exactly what user code would do, also out-of-bounds in terms of when the next request is pulled for... so should be fine I think... Please sanity check my thinking here @raboof

raboof replied:
Yes I agree this looks like it makes sense to me.

ktoso commented on akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala:
```diff
@@ -26,5 +32,84 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
+   * 2) Handle in-flight request:
+   * - if a request is "in-flight" (being handled by user code), it is given `hardDeadline` time to complete,
+   *   - if user code emits a response within the timeout, then this response is sent to the client
+   *     - however if it is a streaming response, it is also mandated that it shall complete within the deadline, and if it does not
```
thanks adding

ktoso commented:
I added better error handling in case draining entities would fail during termination

ktoso commented:
I also thought about bin compat and think we're fine, since it's the 2nd param list, so unapply is not broken. The server binding class is also final, and no one should have ever created instances of it...

ignasi35 commented on akka-http-core/src/test/scala/akka/http/scaladsl/GracefulTerminationSpec.scala:
```diff
@@ -0,0 +1,261 @@
+    }
+  }
+
+}
```
nice!

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -460,7 +465,7 @@ private[http] object HttpServerBluePrint {
         override def onPull() = {
           pull(httpResponseIn)
-          // after the initial pull here we only ever pull after having emitted in `onPush` of `httpResponseIn`
+          // after the iServerTerminationDeadlineReachednitial pull here we only ever pull after having emitted in `onPush` of `httpResponseIn`
```
whoa, thanks for catching~

ktoso commented on akka-http-core/src/main/scala/akka/http/javadsl/ServerBinding.scala:
```diff
@@ -26,5 +32,86 @@ class ServerBinding private[http] (delegate: akka.http.scaladsl.Http.ServerBindi
+   *   yet no new requests will be delivered to the user handler. All such drained responses will be replied to with an
+   *   termination response (as explained in phase 2).
+   *
+   * 3) Close still existing connections
```
thanks!

ktoso commented:
Thanks for review @ignasi35 !

Shall we merge @raboof ?

raboof commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -883,24 +910,126 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
     localAddress:  InetSocketAddress,
     remoteAddress: InetSocketAddress,
-    flow:          Flow[HttpResponse, HttpRequest, NotUsed]) {
+    _flow:         Flow[HttpResponse, HttpRequest, ServerTerminator]) { // FIXME may not change mat value I guess
```
As long as the `NotUsed` instance is not used... or could it lead to a `ClassCastException` somewhere even though the actual value was ignored?

patriknw replied:
FIXME?

ktoso replied:
Removed fixme... I think it could be an issue if someone did mapMaterialized value on a flow obtained by pattern matching i guess... 

Note that for just accessing the flow I kept the `def flow` below so .flow would still be the same type and not cause issues in mapMaterializedValue.

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala:
```diff
@@ -0,0 +1,307 @@
+            // sending the reply here is a "nice to try", but the stage failure will likely overtake it and terminate the connection first
+            emit(toNet, settings.terminationDeadlineExceededResponse, () ⇒ failStage(ex))
+          } else {
+            failStage(ex) // TODO think if to throw or fail here? this causes the "dont keep connections around longer than the deadline"
```
thinking done?

ktoso replied:
Yes, thanks for skimming TODO/FIXMEs, seems I missed some

ktoso replied:
(It is correct to fail stage here)

patriknw commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -588,7 +591,14 @@ abstract class ConnectionPoolSpec(poolImplementation: PoolImplementation) extend
+                // this is simply a mock, since we do not use termination in these tests anyway
+                def terminate(deadline: FiniteDuration)(implicit ec: ExecutionContext): Future[HttpTerminated] =
+                  Future.successful(HttpServerTerminated)
+              }) // FIXME layer should by itself carry the terminator already
```
FIXME?

ktoso commented:
Thanks, sorry I missed those before. Fixed all fixme's (not using calling thread anymore anywhere as well), and replied to bin compat question (i think it's fine, since compat method added and the incompat would be in a crazy case with unapplying the case class and then matmaterializedvalue mapping over the notused etc).

Rebased as well

ktoso commented:
Rebased again, all comments are addressed AFAICS. The TODO that remain are linked to tickets
